exp/exp.go

```diff
@@ -0,0 +1,148 @@
+// Hook go-metrics into expvar
+// on any /debug/vars2 request, load all vars from the registry into expvar, and execute regular expvar handler
+package exp
+
+import (
+	"expvar"
+	"fmt"
+	"github.com/rcrowley/go-metrics"
+	"net/http"
+	"sync"
+)
+
+type exp struct {
+	expvarLock sync.Mutex // expvar panics if you try to register the same var twice, so we must probe it safely
+	registry   metrics.Registry
+}
+
+func (exp *exp) expHandler(w http.ResponseWriter, r *http.Request) {
+	// load our variables into expvar
+	exp.syncToExpvar()
+
+	// now just run the official expvar handler code (which is not publicly callable, so pasted inline)
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	fmt.Fprintf(w, "{\n")
+	first := true
+	expvar.Do(func(kv expvar.KeyValue) {
+		if !first {
+			fmt.Fprintf(w, ",\n")
+		}
+		first = false
+		fmt.Fprintf(w, "%q: %s", kv.Key, kv.Value)
+	})
+	fmt.Fprintf(w, "\n}\n")
+}
+
+func Exp(r metrics.Registry) {
+	e := exp{sync.Mutex{}, r}
+	// this would cause a panic:
+	// panic: http: multiple registrations for /debug/vars
+	// http.HandleFunc("/debug/vars", e.expHandler)
+	// haven't found an elegant way, so just use a different endpoint
+	http.HandleFunc("/debug/vars2", e.expHandler)
```

Comment:
That's a good point. Then how about just 'debug/metrics'? I'm generally not
a fan of adding numbers to variable names and urls where there's an
opportunity to have a meaningful name.

On Saturday, July 4, 2015, Dieter Plaetinck <notifications@github.com>
wrote:

> In exp/exp.go
> <https://github.com/rcrowley/go-metrics/pull/96#discussion_r33890766>:
>
> > +		if !first {
> > +			fmt.Fprintf(w, ",\n")
> > +		}
> > +		first = false
> > +		fmt.Fprintf(w, "%q: %s", kv.Key, kv.Value)
> > +	})
> > +	fmt.Fprintf(w, "\n}\n")
> > +}
> > +
> > +func Exp(r metrics.Registry) {
> > +	e := exp{sync.Mutex{}, r}
> > +	// this would cause a panic:
> > +	// panic: http: multiple registrations for /debug/vars
> > +	// http.HandleFunc("/debug/vars", e.expHandler)
> > +	// haven't found an elegant way, so just use a different endpoint
> > +	http.HandleFunc("/debug/vars2", e.expHandler)
>
> what would be the point of introducing variability for this?
> it makes more sense to me IMHO to have some sort of "convention" for the
> address of a "go-metrics enabled expvar page". one thing is for sure: I
> couldn't find a way to take over/use
> /debug/vars. Actually now that I think about it, /debug/vars is also
> fixed for convention, which I think is a good thing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rcrowley/go-metrics/pull/96/files#r33890766>.
>


-- 
Sent from Russia, with love. But also from my phone.